discv5/discv5-wire.md

```diff
@@ -1,395 +1,307 @@
-  messages will be lost over UDP. Implementations must receive with a
-  timeout.
-- The *conversation nonce* determines the conversation and is used on
-  each message.
-- The precise same message format may be used over TCP, except that
-  the message may include a message "1 of 1" descriptor, include all
-  data in the response stream, and rely on length information
+Node discovery messages are sent as UDP datagrams. Since UDP is a lossy transport, packets
+may be received in any order or not at all. Implementations should not re-send packets if
+the recipient doesn't respond, though there are exceptions to this general rule. If
+multiple requests are pending while performing the handshake, the requests may be re-sent
+with new keys (see [handshake section]). If a node's liveness has been verified many
+times, implementations may consider occasional non-responsiveness permissible and assume
+the node is live.
 
-### General Message Format
-
-**TBD: At the time of writing general consensus seems to be around RLP.
-Still waiting for serious proposals for alternatives.**
-
-Symbols
-
-`[ .. , .. , .. ]` means an RLP list
-`a || b` means concatenation of `a` and `b`
-
-As with v4 messages, a message is:
+The maximum size of any packet is 1280 bytes. Implementations should not generate or
+process packets larger than this size. Most messages are smaller than this limit by
+definition, the exception being the NODES message. FINDNODE returns up to 16 records, plus
+other data, and TOPICQUERY may also distribute a significantly long list of ENRs. As per
+specification the maximum size of an ENR is 300 bytes. A NODES message containing all
+FINDNODE response records would be at least 4800 bytes, not including additional data such
+as the header. To stay below the size limit, NODES responses are sent as multiple messages
+and specify the total number of responses in the message.
 
-```text
-message = message-header || message-data
-```
+Since low-latency communication is expected, implementations should place short timeouts
+on request/response interactions. Good timeout values are 500ms for a single
+request/response and 1s for the handshake.
 
-where
+When responding to a request, the response should be sent to the UDP envelope address of
+the request.
 
-```text
-message-header = hash || signature || message-type
-hash = keccak256(signature || message-type || message-data)
-signature = sign(message-type || message-data)
-```
+Handshake
+---------
+
+Discovery communication is encrypted and authenticated using session keys, established in
+the handshake. Since every node participating in the network acts as both client and
+server, a handshake can be initiated by either side of communication at any time. In the
+following definitions, we assume that node A wishes to communicate with node B, e.g. to
+send a FINDNODE query.
+
+Node A must have a node record for node B and know B's node ID to communicate with it. If
```

Comment:
Not directly related, but one thing I mentioned in the rationale somewhere is that we will want to change the AddPeer RPC and the Bootnodes/static nodes lists so that ENRs will be added and not simple IP/port/ID tuples. This is to make it hard to spawn a node with junk data.